flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -237,57 +230,73 @@ private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerExce
 			++currentMaxAttemptId);
 	}
 
-	private void requestKubernetesPod() {
-		numPendingPodRequests++;
+	private void requestKubernetesPod(WorkerResourceSpec workerResourceSpec) {
+		final KubernetesTaskManagerParameters parameters =
+			createKubernetesTaskManagerParameters(workerResourceSpec);
+
+		final KubernetesPod taskManagerPod =
+			KubernetesTaskManagerFactory.createTaskManagerComponent(parameters);
+		kubeClient.createTaskManagerPod(taskManagerPod);
+
+		podWorkerResources.put(parameters.getPodName(), workerResourceSpec);
+		final int pendingWorkerNum = notifyNewWorkerRequested(workerResourceSpec);
 
 		log.info("Requesting new TaskManager pod with <{},{}>. Number pending requests {}.",
-			defaultMemoryMB,
-			defaultCpus,
-			numPendingPodRequests);
+			parameters.getTaskManagerMemoryMB(),
+			parameters.getTaskManagerCPU(),
+			pendingWorkerNum);
+		log.info("TaskManager {} will be started with {}.", parameters.getPodName(), workerResourceSpec);
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(WorkerResourceSpec workerResourceSpec) {
+		final TaskExecutorProcessSpec taskExecutorProcessSpec =
+			TaskExecutorProcessUtils.processSpecFromWorkerResourceSpec(flinkConfig, workerResourceSpec);
 
 		final String podName = String.format(
 			TASK_MANAGER_POD_FORMAT,
 			clusterId,
 			currentMaxAttemptId,
 			++currentMaxPodId);
 
+		final ContaineredTaskManagerParameters taskManagerParameters =
+			ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorProcessSpec);
+
 		final String dynamicProperties =
 			BootstrapTools.getDynamicPropertiesAsString(flinkClientConfig, flinkConfig);
 
-		final KubernetesTaskManagerParameters kubernetesTaskManagerParameters = new KubernetesTaskManagerParameters(
+		return new KubernetesTaskManagerParameters(
 			flinkConfig,
 			podName,
 			dynamicProperties,
 			taskManagerParameters);
-
-		final KubernetesPod taskManagerPod =
-			KubernetesTaskManagerFactory.createTaskManagerComponent(kubernetesTaskManagerParameters);
-
-		log.info("TaskManager {} will be started with {}.", podName, taskExecutorProcessSpec);
-		kubeClient.createTaskManagerPod(taskManagerPod);
 	}
 
 	/**
 	 * Request new pod if pending pods cannot satisfy pending slot requests.
 	 */
-	private void requestKubernetesPodIfRequired() {
-		final int requiredTaskManagers = getNumberRequiredTaskManagers();
+	private void requestKubernetesPodIfRequired(WorkerResourceSpec workerResourceSpec) {
+		final int pendingWorkerNum = getNumPendingWorkersFor(workerResourceSpec);
+		int requiredTaskManagers = getRequiredResources().get(workerResourceSpec);
 
-		while (requiredTaskManagers > numPendingPodRequests) {
-			requestKubernetesPod();
+		while (requiredTaskManagers-- > pendingWorkerNum) {
+			requestKubernetesPod(workerResourceSpec);
 		}
 	}
 
 	private void removePodIfTerminated(KubernetesPod pod) {
 		if (pod.isTerminated()) {
 			kubeClient.stopPod(pod.getName());
```

Comment:
Can it happen that we stop a pod w/o running `onAdded` which decrease the super classes `PendingWorkerCounter`? If yes, then the internal state might go out of sync.

Reply:
Nope. There are two occurrence of `kubeClient.stopPod` and they're both guaranteed to be called after `onAdded`.

Reply:
So this means that `KubernetesResourceManager.onError` will only be called if `onAdded` has been called before? I guess this is also a question for @wangyang0918.

Reply:
I post the `WatchEvent` api in K8s document[1]. From the definition, i do not find the the guarantee that `Error` event is always returned after `Added`. We already have a [ticket](https://issues.apache.org/jira/browse/FLINK-17177) to track how to handle the `onError` event in `KubernetesResourceManager` correctly. I think we could keep the current behavior and try to find a solution in FLINK-17177.

Moerover, i need to dig into the K8s ApiServer implementation and check when the `onError` `WacherEvent` could be returned.


[1]. https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#watchevent-v1-meta

Reply:
> So this means that `KubernetesResourceManager.onError` will only be called if `onAdded` has been called before? I guess this is also a question for @wangyang0918.

There is no guarantee for this, `ERROR` could be thrown in https://github.com/kubernetes/kubernetes/blob/343c1e7636fe5c75cdd378c0b170b26935806de5/staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go#L121

Reply:
It seems I was wrong about `onError` always called after `onAdded`.
Then isn't `numPendingPodRequests` already possibly go out of sync before this PR? Do we observe any cases?

@zhengcanbin Would it be possible that a pod is never successfully started, that we received `onError` but never `onAdded`, `onModified` or `onDeleted`?

Reply:
> Then isn't `numPendingPodRequests` already possibly go out of sync before this PR? Do we observe any cases?

Yes, I noticed this problem when I dived into `KubernetesResourceManager` yesterday, it is possible that we do not receive an `ADD` Event(we met this kind of problem in our K8s production env, but not in Flink so far) even if the Pod has been deployed successfully on the K8s cluster.

> @zhengcanbin Would it be possible that a pod is never successfully started, that we received `onError` but never `onAdded`, `onModified` or `onDeleted`?

It is possible if something goes wrong in the HTTP stream and the K8s server sends an `ERROR`, but I think this should rarely happen.

Reply:
It seems there are two issues.
1. `numPendingPodRequests`/`PendingWorkerCounter` may go out of sync, if pod is stopped in `onError` before `onAdded` is called.
2. Whether should we remove the pod or not when `onError` is called.

I would suggest the following.
- Try to resolve (1) in this PR, based on the current behavior that pod will be removed when `onError` is called. We can check whether the pod is requested in the current attempt by looking up in `podWorkerResources`, and whether the `onAdded` has been called by looking up in `workerNodes`. If the pod is requested in the current attempt and `onAdded` is not called yet, then we should decrease the pending count.
- Leave (2) to be discussed and resolved in FLINK-17177. We may or may not need to remove the pod when `onError` is called. Moreover, if we remove the pod and decrease the pending count, should we try to request another new pod? To answer these questions, I guess we need to better understand under what circumstances will an `ERROR` be received.

WDYT? @wangyang0918 @zhengcanbin @tillrohrmann

Reply:
I agree with @xintongsong to make sure the `numPendingPodRequests/PendingWorkerCounter` are correctly increased and decreased in this PR. If the `onError` happens before `onAdded`, then we should also decrease the `numPendingPodRequests`.

For how to handle `onError`, let's keep the discussion in FLINK-17177. Since i find the fabric8 kubernetes client has a different implementation with K8s go client. It will never produce a `onError` `WatchEvent` on client side[1]. But i am not familiar with the K8s apiserver about when it will return a `Error` type `WatchEvent`.

[1]. https://github.com/fabric8io/kubernetes-client/blob/1afea0c364a48d9a8745539f84de59630ef6f559/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java#L289

Reply:
I think it is a good idea to split these two efforts. 

In fact for (1), I think we need the same checks for `onModified` and `onDeleted` as well. Hence, `removePodIfTerminated` should check whether it is an intended shutdown or not (imagine a manual termination of a pod). In case that it is not, then it needs to decrease the `PendingWorkerCounter` and call `requestKubernetesPodIfRequired`.

It looks as if `removePodIfTerminated` already does this in the case that a pod has been started (`onAdded` has been called) and if the pod fails then. However, it does not do it in case `onAdded` has not been called. But in this case, I think we should request a new pod or fail fatally.

Reply:
I think it could work by using the `podWorkerResources` map to indicate whether a pod needs to be restarted or not. Whenever it contains an entry for a given `ResourceID` we need to restart it. This, of course, means that we need to clear this entry in `stopWorker`.

Reply:
I'm actually thinking about similar approach. I think we should do two things for all the stopped / terminated workers.
- Check whether `onAdded` is called by looking into `workerNodes`. If it's not called, we should decrease the pending count.
- Try to restart the pod if required. As for indicating whether to restart the worker or not, I'm in favor of relying on `getRequiredResources` which always fetch the up-to-date number of required resources from `SlotManager`, rather than relying on `podWorkerResources`. But ideally these two should be consistent if we properly clean-up `podWorkerResources`.

I'm trying to rebase this PR onto the latest master for conflict resolving. And there's a newly introduced `internalStopPod` which might be a good place for applying the above two actions for all stopped / terminated workers.

Reply:
I'd also be fine with solving this problem in a separate effort. I think the problem has been part of the `KubernetesResourceManager` before. My suggestion for this effort would be to add thorough test coverage. Moreover, it might make sense to see whether there is similar functionality in the `YarnResourceManager` and to think whether it can/should be moved to the `ActiveResourceManager`. I have the feeling that we are re-implementing logic for different `ResourceManager` implementations.

Reply:
I am afraid that the situation is different in YARN. In `YarnResourceManager`, when the `AMRMClient` heartbeats failed, may because YARN RM internal error, or invalid resource request, or network problems, or sth else, we will call `onFatalError` to fail the current jobmanager attempt and start a new one. We do not retry in such case.

Maybe we could have the same logics in `KubernetesResourceManager` if we confirm that the `Error WatchEvent` is sent only when K8s ApiServer internal error, or invalid pod spec. I would give more information in FLINK-17177 after dive into the K8s ApiServer implementation.

Till is right, we need to add thorough test coverage for this effort and try to move the same logics to `ActiveResouceManager`.

Reply:
@xintongsong Sorry for the late reply. I accidentally jumped into this PR a few days ago and currently, I haven't taken a closer look at this PR. But recently I have two concerns:
1. Maybe our handle logic should not rely on the `EVENT` order since there is no guarantee for this, also, the `EVENT` from the `Watcher` stream may not have exactly-once semantics, we may lose some of them or receive the same `EVNET` more than once.
2. We rely heavily on the informer of `Watcher` to track the state of the Pods, I am not entirely sure that this is a good solution, maybe we need an active polling mechanism as a supplement to get the whole view of the Pods. Of course, it needs further discussion and could be resolved in another ticket if necessary, which is out of the scope of this PR.

Reply:
Thanks all for the feedbacks and good discussion.
@zhengcanbin, I've added more checks to avoid relying on the order and non-duplication of received events.
@tillrohrmann, I've also added more test cases verifying the pending worker counter with out-of-ordered and duplicated events.